File: app/src/main/java/ir/ghararemaghzha/game/dialogs/UserDetailsDialog.java
package ir.ghararemaghzha.game.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.bumptech.glide.Glide;
import com.google.android.material.textview.MaterialTextView;
import com.willy.ratingbar.ScaleRatingBar;

import ir.ghararemaghzha.game.R;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.data.RetrofitClient;
import ir.ghararemaghzha.game.models.UserDetailsResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserDetailsDialog extends Dialog {

    private final Activity context;
    private final String userId;
    private MaterialTextView name, code, rank, score, questions, answers, questionsPercent, answersPercent, rateText;
    private ProgressBar answersProgress, questionsProgress;
    private ImageView avatar;
    private ScaleRatingBar ratingBar;

    private ConstraintLayout loading;

    public UserDetailsDialog(@NonNull Activity context, String userId) {
        super(context);
        this.context = context;
        this.userId = userId;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.dialog_userdetails);
        init();
    }

    private void init() {
        loading = findViewById(R.id.details_loading);

        ratingBar = findViewById(R.id.details_rate);
        ratingBar.setRating(0f);
        ratingBar.setClickable(false);
        ratingBar.setScrollable(false);
        ratingBar.setStarPadding(8);
        rateText = findViewById(R.id.details_total);
        name = findViewById(R.id.details_name2);
        code = findViewById(R.id.details_code);
        rank = findViewById(R.id.details_rank2);
        score = findViewById(R.id.details_rate_text);
        questions = findViewById(R.id.details_questions);
        questionsPercent = findViewById(R.id.details_questions_percent);
        answers = findViewById(R.id.details_answers);
        answersPercent = findViewById(R.id.details_answers_percent);

        avatar = findViewById(R.id.details_avatar);

        answersProgress = findViewById(R.id.details_answers_progress);
        answersProgress.setProgress(0);
        answersProgress.setMax(100);

        questionsProgress = findViewById(R.id.details_questions_progress);
        questionsProgress.setProgress(0);
        questionsProgress.setMax(100);

        findViewById(R.id.details_close).setOnClickListener(v -> dismiss());

        getData();
    }

    private void getData() {
        String number = MySharedPreference.getInstance(context).getNumber();
        String token = MySharedPreference.getInstance(context).getAccessToken();
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(context, true);
            return;
        }
        RetrofitClient.Companion.getInstance().getApi()
                .getUserDetails("Bearer " + token, number, userId)
                .enqueue(new Callback<UserDetailsResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<UserDetailsResponse> call, @NonNull Response<UserDetailsResponse> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {
                            if (response.body().getMessage().equals("ok")) {
                                name.setText(response.body().getUserData().getUserName());
                                code.setText(context.getString(R.string.details_code, response.body().getUserData().getUserCode()));
                                rank.setText(response.body().getUserData().getUserRank());
                                score.setText(response.body().getUserData().getScoreCount());
                                if (Integer.parseInt(response.body().getBooster()) < 1500)
                                    answers.setText(context.getString(R.string.details_nitro2, response.body().getBooster()));
                                else
                                    answers.setText(context.getString(R.string.details_nitro2, "1500"));

                                Glide.with(context)
                                        .load(context.getString(R.string.avatar_url, response.body().getUserData().getUserAvatar()))
                                        .circleCrop()
                                        .placeholder(R.drawable.placeholder)
                                        .into(avatar);

                                int totalQuestions = (Integer.parseInt(response.body().getPlan()) * 500) + 500;
                                int answeredQuestions = response.body().getCorrect() + response.body().getIncorrect();
                                int nitroUsed = Integer.parseInt(response.body().getBooster());

                                if (answeredQuestions <= 3000) {
                                    questions.setText(context.getString(R.string.details_questions, answeredQuestions, totalQuestions));
                                } else {
                                    questions.setText(context.getString(R.string.details_questions, 3000, totalQuestions));
                                }

                                int qPercent;
                                int nPercent;
                                if (answeredQuestions > 0 && answeredQuestions <= 3000) {
                                    qPercent = (answeredQuestions * 100) / totalQuestions;
                                    nPercent = (nitroUsed * 100) / 1500;
                                } else if (answeredQuestions > 3000) {
                                    qPercent = (answeredQuestions * 100) / totalQuestions;
                                    nPercent = (nitroUsed * 100) / 1500;
                                } else {
                                    qPercent = 0;
                                    nPercent = 0;
                                }
                                new Handler().postDelayed(() -> {
                                    if (Build.VERSION.SDK_INT >= 24) {
                                        answersProgress.setProgress(nPercent, true);
                                        questionsProgress.setProgress(qPercent, true);
                                    } else {
                                        answersProgress.setProgress(nPercent);
                                        questionsProgress.setProgress(qPercent);
                                    }
                                }, 500);

                                answersPercent.setText("%" + nPercent);

                                questionsPercent.setText("%" + qPercent);

                                int rate = Integer.parseInt(response.body().getRank());
                                int level1 = Integer.parseInt(response.body().getLevel1());
                                int level2 = Integer.parseInt(response.body().getLevel2());
                                int level3 = Integer.parseInt(response.body().getLevel3());
                                int level4 = Integer.parseInt(response.body().getLevel4());
                                rateText.setText( String.valueOf(rate));
                                if (rate < level1) {
                                    ratingBar.setRating(0f);
                                } else if (rate < level2)
                                    ratingBar.setRating(1f);
                                else if (rate < level3)
                                    ratingBar.setRating(2f);
                                else if (rate < level4)
                                    ratingBar.setRating(3f);
                                else
                                    ratingBar.setRating(4f);


                                loading.setVisibility(View.GONE);
                            } else {
                                Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                                dismiss();
                            }
                        } else if (response.code() == 401) {
                            dismiss();
                            Utils.logout(context, true);
                        } else {
                            Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                            dismiss();
                        }

                    }

                    @Override
                    public void onFailure(@NonNull Call<UserDetailsResponse> call, @NonNull Throwable t) {
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                        dismiss();
                    }
                });


    }

    @Override
    public void dismiss() {
        answersProgress.setProgress(0);
        questionsProgress.setProgress(0);
        super.dismiss();

    }
}

File: app/src/main/java/ir/ghararemaghzha/game/dialogs/UserDetailsDialog.kt
package ir.ghararemaghzha.game.dialogs

import android.app.Activity
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import com.bumptech.glide.Glide
import ir.ghararemaghzha.game.R
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.data.ApiRepository
import ir.ghararemaghzha.game.data.NetworkApi
import ir.ghararemaghzha.game.data.RemoteDataSource
import ir.ghararemaghzha.game.data.Resource
import ir.ghararemaghzha.game.databinding.DialogUserdetailsBinding
import ir.ghararemaghzha.game.viewmodels.UserDetailsViewModel
import ir.ghararemaghzha.game.viewmodels.ViewModelFactory
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class UserDetailsDialog(ctx: FragmentActivity, private val userId: String) : Dialog(ctx) {

    private lateinit var b: DialogUserdetailsBinding
    private lateinit var number: String
    private lateinit var token: String
    private lateinit var viewModel: UserDetailsViewModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        b = DialogUserdetailsBinding.inflate(LayoutInflater.from(context))
        val factory = ViewModelFactory(ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java)))
        //  viewModel = ViewModelProvider(ctx, factory).get(UserDetailsViewModel::class.java)
        viewModel = factory.create(UserDetailsViewModel::class.java)
        setContentView(b.root)

        init()
    }


    private fun init() {
        getUserDetails()
        b.detailsRate.rating = 0f
        b.detailsRate.isClickable = false
        b.detailsRate.isScrollable = false
        b.detailsRate.starPadding = 8
        b.detailsAnswersProgress.progress = 0
        b.detailsAnswersProgress.max = 100
        b.detailsQuestionsProgress.progress = 0
        b.detailsQuestionsProgress.max = 100
        b.detailsClose.setOnClickListener { dismiss() }
        b.detailsLoading.visibility = View.VISIBLE

        //  viewModel.getUserDetails("Bearer $token", number, userId)
        CoroutineScope(Dispatchers.IO).launch {
            getData()
        }
    }

    override fun dismiss() {
        super.dismiss()
        b.detailsAnswersProgress.progress = 0
        b.detailsQuestionsProgress.progress = 0
    }

    private fun getUserDetails() {
        number = MySharedPreference.getInstance(context).getNumber()
        token = MySharedPreference.getInstance(context).getAccessToken()
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(context as Activity, true)
        }
    }

    private suspend fun getData() {
        when (val res = ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java)).getUserDetails("Bearer $token", number, userId)) {
            is Resource.Success -> {
                if (res.value.result == "success" && res.value.message == "ok") {
                    withContext(Dispatchers.Main) {
                        b.detailsName2.text = res.value.userData.userName
                        b.detailsCode.text = context.getString(R.string.details_code, res.value.userData.userCode)
                        b.detailsRank2.text = res.value.userData.userRank

                        if (res.value.userData.scoreCount != "-1")
                            b.detailsRateText.text = res.value.userData.scoreCount
                        else
                            b.detailsRateText.text = "0"

                        if (res.value.booster.toInt() < 1000) b.detailsAnswers.text = context.getString(R.string.details_nitro2, res.value.booster) else b.detailsAnswers.text = context.getString(R.string.details_nitro2, "1000")

                        Glide.with(context)
                                .load(context.getString(R.string.avatar_url, res.value.userData.userAvatar))
                                .circleCrop()
                                .placeholder(R.drawable.placeholder)
                                .into(b.detailsAvatar)

                        val totalQuestions = 1000
                        val answeredQuestions: Int = res.value.correct + res.value.incorrect

                        val nitroUsed = if (res.value.booster.toInt() > 1000)
                            1000
                        else
                            res.value.booster.toInt()

                        if (answeredQuestions <= 1000) {
                            b.detailsQuestions.text = context.getString(R.string.details_questions, answeredQuestions, totalQuestions)
                        } else {
                            b.detailsQuestions.text = context.getString(R.string.details_questions, 1000, totalQuestions)
                        }

                        val qPercent: Int
                        val nPercent: Int
                        when {
                            answeredQuestions in 1..1000 -> {
                                qPercent = answeredQuestions * 100 / totalQuestions
                                nPercent = nitroUsed * 100 / 1000
                            }
                            answeredQuestions > 1000 -> {
                                qPercent = 100
                                nPercent = nitroUsed * 100 / 1000
                            }
                            else -> {
                                qPercent = 0
                                nPercent = 0
                            }
                        }

                        b.detailsAnswersProgress.progress = nPercent
                        b.detailsQuestionsProgress.progress = qPercent


                        b.detailsAnswersPercent.text = "%$nPercent"
                        b.detailsQuestionsPercent.text = "%$qPercent"

                        val rate = res.value.rank.toInt()
                        val level1 = res.value.level1.toInt()
                        val level2 = res.value.level2.toInt()
                        val level3 = res.value.level3.toInt()
                        val level4 = res.value.level4.toInt()
                        b.detailsTotal.text = rate.toString()
                        when {
                            rate < level1 -> b.detailsRate.rating = 0f
                            rate < level2 -> b.detailsRate.rating = 1f
                            rate < level3 -> b.detailsRate.rating = 2f
                            rate < level4 -> b.detailsRate.rating = 3f
                            else -> b.detailsRate.rating = 4f
                        }
                        b.detailsLoading.visibility = View.GONE
                    }

                } else {
                    withContext(Dispatchers.Main) {
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show()
                        dismiss()
                    }
                }
            }
            is Resource.Failure -> {
                if (res.isNetworkError) {
                    withContext(Dispatchers.Main) {
                        b.detailsLoading.visibility = View.GONE
                        Toast.makeText(context, context.getString(R.string.general_error), Toast.LENGTH_SHORT).show()
                        dismiss()
                    }
                } else if (res.errorCode == 401) {
                    withContext(Dispatchers.Main) {
                        b.detailsLoading.visibility = View.GONE
                        Utils.logout(context as Activity, true)
                    }
                }
            }
            is Resource.Loading -> {
            }
        }
    }
}